ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -372,10 +381,15 @@ short getResetKeyLifeVersion() throws StoreException {
    */
   NavigableSet<IndexValue> find(StoreKey keyToFind) throws StoreException {
     NavigableSet<IndexValue> toReturn = null;
-    rwLock.readLock().lock();
+    // We deliberately get a reference to index before calling isSealedLocked.
+    // With rwLock, "serEntries != null" is an atomic operation with "sealed.get() == true", which means,
+    // when sealed.get() == true, then serEntries will not be null.
+    // Also, getting reference of index before checking sealed.get() would make sure tha when sealed.get()
+    // returns false, the reference will always be not null.
+    NavigableMap<StoreKey, ConcurrentSkipListSet<IndexValue>> indexCopy = index;
     try {
-      if (!sealed.get()) {
-        ConcurrentSkipListSet<IndexValue> values = index.get(keyToFind);
+      if (!isSealedLocked()) {
+        ConcurrentSkipListSet<IndexValue> values = indexCopy.get(keyToFind);
```

Comment:
Once isSealedLocked() returns, no more lock is held. What if it returns false and just before you are executing the else branch it is set to true by another thread.

We need to get and keep the lock within the whole try block.

Reply:
I have different idea about the lock here. It's actually fine to not hold the lock through out the entire method.
First, all the changes are done in read-only methods, so there will be no concurrent writes here. 
Secondly, sealed or unsealed, IndexSegment has the same information, just managed in different way. For example, this find method, if after `isSealedLock()` returns false, and other thread set it to true right way, we can still use `indexCopy` map to find the index values. Same logic applies to other methods.

Reply:
I am a little confused. If this is a single-threaded execution, why do we need to protect the "sealed" in the first place?

For the second part:
  - If the critical path does not involve the whole routine but just the subset of it, you can reduce the scope of the lock.
  - If segment's being locked or not does not make any difference:
    + Can we guarantee that it will always be true in the future? 
    + Can we consider having a single code path?

Reply:
this is not a single-thread execution of course. When an IndexSegment is created, it is unsealed. It has a map from blob id to index entries. This map is the variable `index`. After it becomes sealed, this map will be persisted to disk and loaded back to memory in a `ByteBuffer`. This `ByteBuffer` is the variable `serEntries`. 

The thread that changes index segment from unsealed to sealed is a background thread and we also have multiple read threads that returns index values from a given blob id. 

And checking if the index segment is sealed or not will help us determine which variable to use when returning index values. If index segment is sealed, then we should use `serEntries` buffer, if not, we should use `index` map. So as long as we know which variable to use, we can return the index values from this variable without locking, because if it's sealed, buffer `serEntries` is a read only buffer. if it's not sealed, `index` map is a concurrent hash map that can handles concurrent reads and writes.

Reply:
Thanks for the explanation.

Since you make a copy of `index` before calling isSealedLocked(), this would work. Having said that it is not a good practice in my opinion. Ideally, all the critical code path needs to be executed with the read-lock in order to be future-proof.

I don't expect you to make the changes in this PR but just noting the potential issue going forward.

Reply:
No urgency for this PR, I will update all.